Intermittent MaxRetryError (and others) in CI unit test runs

**Describe the bug**

Recently, CI unit test suite jobs have been producing increasingly frequent intermittent test failures due to time-outs and excessive retries. The problem seems to occur most frequently on GitHub, but occasionally also on AppVeyor. This issue was prompted by [this GitHub actions run](https://github.com/dcs4cop/xcube/actions/runs/6340782131/job/17222954166) which produced the following error:

```
FAILED test/webapi/ows/stac/test_routes.py::StacRoutesTest::test_fetch_catalog_collection_single_items - 
urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=34113): Max retries exceeded with url: 
/ogc/collections/demo/items (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f8a94db71d0>: 
Failed to establish a new connection: [Errno 111] Connection refused'))
```

We can use this issue to collect further instances of the problem.

So far our fix for such errors has been "re-run the job and hope it goes away", which it generally does, but this is turning into something of a time sink.

**To Reproduce**
Steps to reproduce the behavior: keep re-running the GitHub unittest job until an error occurs.

**Expected behavior**
All tests should pass reliably on every CI run.

**Additional context**
With luck, this might be fairly easily fixable by tweaking some back-off / time-out / retry parameters in `ServerTestCase` or similar.

Another one, from <https://github.com/dcs4cop/xcube/actions/runs/6467629870/job/17559781773>:

```text
FAILED test/webapi/places/test_routes.py::PlacesRoutesTest::test_places - urllib3.exceptions.MaxRetryError: 
HTTPConnectionPool(host='localhost', port=33397): Max retries exceeded with url: /places (Caused by 
NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7f6942170a10>: Failed to establish a new connection: 
[Errno 111] Connection refused'))
```

<https://github.com/dcs4cop/xcube/actions/runs/6467629870/job/17571143347>

```text
FAILED test/webapi/datasets/test_routes.py::DatasetsRoutesTest::test_fetch_datasets - urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=39147): Max retries exceeded with url: /datasets (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb970811990>: Failed to establish a new connection: [Errno 111] Connection refused'))
FAILED test/webapi/s3/test_routes.py::S3RoutesNewTest::test_fetch_get_s3_object - urllib3.exceptions.MaxRetryError: HTTPConnectionPool(host='localhost', port=54325): Max retries exceeded with url: /s3/datasets/demo.zarr/.zattrs (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x7fb98c58f950>: Failed to establish a new connection: [Errno 111] Connection refused'))
```

--- a/xcube/server/testing.py
+++ b/xcube/server/testing.py
@@ -31,6 +31,7 @@ from typing import Dict, Optional, Any, Union, Tuple
 
 import urllib3
 from tornado.platform.asyncio import AnyThreadEventLoopPolicy
+from urllib3.exceptions import MaxRetryError
 
 from xcube.server.server import Server
 from xcube.server.webservers.tornado import TornadoFramework
@@ -66,13 +67,20 @@ class ServerTestCase(unittest.TestCase, ABC):
 
         self.http = urllib3.PoolManager()
 
+        # Taking care that server is fully started until tests make requests.
+        # Fixing https://github.com/dcs4cop/xcube/issues/899
+        elapsed_time = 0.0
         while True:
-            # noinspection PyBroadException
+            if elapsed_time > 60:
+                self.fail('Server did not respond after one minute. '
+                          'Failing the test.')
             try:
                 self.fetch('/')
                 break
-            except Exception:
-                time.sleep(0.1)
+            except (MaxRetryError, TimeoutError):
+                sleep_time = 0.1
+                time.sleep(sleep_time)
+                elapsed_time += sleep_time
 
     def tearDown(self) -> None:
         self.server.stop()

--- a/xcube/server/testing.py
+++ b/xcube/server/testing.py
@@ -23,6 +23,7 @@ import asyncio
 import json
 import socket
 import threading
+import time
 import unittest
 from abc import ABC
 from contextlib import closing
@@ -64,6 +65,14 @@ class ServerTestCase(unittest.TestCase, ABC):
         tornado.start()
 
         self.http = urllib3.PoolManager()
+
+        while True:
+            # noinspection PyBroadException
+            try:
+                self.fetch('/')
+                break
+            except Exception:
+                time.sleep(0.1)
 
     def tearDown(self) -> None:
         self.server.stop()

--- a/xcube/server/testing.py
+++ b/xcube/server/testing.py
@@ -69,18 +69,12 @@ class ServerTestCase(unittest.TestCase, ABC):
 
         # Taking care that server is fully started until tests make requests.
         # Fixing https://github.com/dcs4cop/xcube/issues/899
-        elapsed_time = 0.0
         while True:
-            if elapsed_time > 60:
-                self.fail('Server did not respond after one minute. '
-                          'Failing the test.')
             try:
                 self.fetch('/')
                 break
             except (MaxRetryError, TimeoutError):
-                sleep_time = 0.1
-                time.sleep(sleep_time)
-                elapsed_time += sleep_time
+                time.sleep(0.1)
 
     def tearDown(self) -> None:
         self.server.stop()

--- a/xcube/server/testing.py
+++ b/xcube/server/testing.py
@@ -67,7 +67,7 @@ class ServerTestCase(unittest.TestCase, ABC):
 
         self.http = urllib3.PoolManager()
 
-        # Taking care that server is fully started until tests make requests.
+        # Taking care that server is fully started before tests make requests.
         # Fixing https://github.com/dcs4cop/xcube/issues/899
         while True:
             try:

